### PR TITLE
PERF/CLN: Use Numpy C Iterator API to unify isnaobj 1D/2D

### PR DESCRIPTION
**Pull request by lithomas1**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

ASV's
```
       before           after         ratio
     [d3995f76]       [aff44731]
     <main>           <clean-perf-isna>
-         104±5ms       31.0±0.3ms     0.30  frame_methods.Isnull.time_isnull_strngs
-         107±7ms         30.5±2ms     0.29  frame_methods.Isnull.time_isnull_obj

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

### CODE REVIEW

File: pandas/_libs/missing.pyx [hunk 1]
```diff
@@ -4,10 +4,16 @@ from sys import maxsize
 
 cimport cython
 from cython cimport Py_ssize_t
+
 import numpy as np
 
 cimport numpy as cnp
 from numpy cimport (
+    PyArray_GETITEM,
```
Comment (jbrockmendel) [inline pandas/_libs/missing.pyx:12]: elsewhere we do "cnp.PyArray_FOO" instead of importing these

File: pandas/_libs/missing.pyx [hunk 2]
```diff
@@ -197,56 +203,22 @@ cpdef ndarray[uint8_t] isnaobj(ndarray arr, bint inf_as_na=False):
...
-    """
-    cdef:
-        Py_ssize_t i, j, n, m
-        object val
-        ndarray[uint8_t, ndim=2] result
-
-    assert arr.ndim == 2, "'arr' must be 2-D."
+        bint is_null
+        ndarray result = np.empty((<object>arr).shape, dtype=np.uint8)
+        flatiter it = PyArray_IterNew(arr)
```
Comment (jbrockmendel) [inline pandas/_libs/missing.pyx:210]: is this robust to F-major inputs?  i think we use the cnp.broadcast pattern for 2D support elsewhere

Comment (lithomas1) [reply pandas/_libs/missing.pyx:210]: I think ``PyArray_IterNew`` forces C style iteration.
https://numpy.org/devdocs/reference/c-api/array.html#c.PyArray_IterNew

Comment (jbrockmendel) [reply pandas/_libs/missing.pyx:210]: oh neat.  then it may turn out some places where ive used cnp.broadcast may be unnecessary

**Review by mroeschke — COMMENTED** [review]
Looks good just a merge conflict

**Comment by mroeschke** [comment]
```
mypy (stubtest)..........................................................Failed
- hook id: stubtest
- duration: 94.59s
- exit code: 1

pandas._libs.missing.isnaobj2d is not present at runtime
```

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @lithomas1

**Pull request merged**